File: frontend/src/types/technical-analysis.ts
/**
 * TypeScript interfaces for Technical Analysis data
 */

export interface TechnicalAnalysisData {
  ticker: string
  analysis_date: string
  current_price: number

  // Trend Analysis
  trend_direction: 'bullish' | 'bearish' | 'neutral'
  trend_strength_score: number
  sma_20: number
  sma_50: number
  sma_200: number
  adx: number
  adx_signal: 'weak' | 'moderate' | 'strong' | 'very_strong'
  price_above_sma_20: boolean
  price_above_sma_50: boolean
  price_above_sma_200: boolean
  golden_cross: boolean
  death_cross: boolean

  // Momentum Analysis
  rsi: number
  rsi_signal: 'oversold' | 'neutral' | 'overbought'
  rsi_weighted_signal: string  // Trend-context adjusted signal
  rsi_weight: number  // Weight applied based on trend context
  macd: number
  macd_signal: number
  macd_histogram: number
  macd_cross?: 'bullish' | 'bearish' | null
  stoch_k: number
  stoch_d: number
  stoch_signal: 'oversold' | 'neutral' | 'overbought'
  roc: number
  roc_signal: 'bullish' | 'neutral' | 'bearish'
  momentum_score: number

  // Volatility Analysis
  bb_upper: number
  bb_middle: number
  bb_lower: number
  bb_width: number
  bb_signal: 'squeeze' | 'breakout_upper' | 'breakout_lower' | 'neutral'
  price_position: 'above_upper' | 'upper' | 'middle' | 'lower' | 'below_lower'
  atr: number
  atr_percent: number
  volatility_level: 'low' | 'moderate' | 'high' | 'very_high'
  volatility_score: number

  // Volume Analysis
  current_volume: number
  avg_volume_20d: number
  volume_ratio: number
  volume_signal: 'low' | 'normal' | 'high' | 'very_high'
  obv: number
  obv_trend: 'rising' | 'falling' | 'neutral'
  volume_score: number

  // Support/Resistance
  pivot: number
  resistance_1: number
  resistance_2: number
  resistance_3: number
  support_1: number
  support_2: number
  support_3: number
  support_levels: number[]
  resistance_levels: number[]
  nearest_support?: number | null
  nearest_resistance?: number | null
  support_distance_pct: number
  resistance_distance_pct: number

  // Chart Patterns
  patterns: string[]
  trend_channel?: 'ascending' | 'descending' | 'horizontal' | null
  consolidation: boolean
  breakout_signal?: 'bullish' | 'bearish' | null

  // Overall Scoring
  trend_score: number
  price_action_score?: number  // Support/Resistance based entry quality score (0-10)
  composite_technical_score: number
  overall_signal: 'strong_buy' | 'buy' | 'neutral' | 'sell' | 'strong_sell'
  signal_confidence: number

  // Multi-timeframe Analysis
  multi_timeframe: MultiTimeframeAnalysis

  // Beta Analysis
  beta_analysis: BetaAnalysis

  // Entry Analysis (comprehensive entry point evaluation)
  entry_analysis?: EntryAnalysis

  // Chart Data
  chart_data: ChartData

  // Data Sources
  data_sources: Record<string, { type: string; name: string }>
}

export interface TimeframeAnalysis {
  timeframe: string  // "daily", "60min", "5min"
  trend_direction: 'bullish' | 'bearish' | 'neutral'
  trend_strength: number
  ema_200_trend: string
  momentum_signal: string
  entry_signal?: string | null
}

export interface MultiTimeframeAnalysis {
  primary_trend: TimeframeAnalysis | null  // Daily
  confirmation_trend: TimeframeAnalysis | null  // 60-min
  execution_trend: TimeframeAnalysis | null  // 5-min
  trend_alignment: 'aligned_bullish' | 'aligned_bearish' | 'mixed'
  signal_quality: 'high' | 'medium' | 'low'
  recommended_action: string
  confidence: number
}

export interface BetaAnalysis {
  beta: number
  benchmark: string
  correlation: number
  alpha: number
  r_squared: number
  volatility_vs_market: 'low' | 'below_average' | 'average' | 'above_average' | 'high'
  risk_profile: 'conservative' | 'moderate' | 'aggressive' | 'very_aggressive'
}

export interface EntryAnalysis {
  // Range Position (0-100%): 0% = at support, 100% = at resistance
  range_position_pct: number
  range_position_zone: 'discount' | 'neutral' | 'premium'

  // Confluence Score (0-10): Weighted sum of confirming factors
  confluence_score: number
  confluence_factors: string[]

  // Stop-Loss Levels
  suggested_stop_loss: number
  stop_loss_type: 'support' | 'atr' | 'swing_low'
  stop_loss_distance_pct: number

  // Take Profit / Target
  suggested_target: number
  target_distance_pct: number

  // Risk/Reward Calculation
  risk_reward_ratio: number
  risk_reward_quality: 'excellent' | 'good' | 'acceptable' | 'poor'

  // Entry Quality Assessment
  is_good_entry: boolean
  entry_quality: 'excellent' | 'good' | 'acceptable' | 'poor'
  entry_quality_score: number  // 0-100

  // Suggested Entry Points
  suggested_entry_price: number
  suggested_entry_zone_low: number
  suggested_entry_zone_high: number
  wait_for_pullback: boolean

  // Analysis Reasoning
  entry_reasoning: string
  warning_signals: string[]
}

export interface ChartData {
  dates: string[]
  ohlcv: {
    open: number[]
    high: number[]
    low: number[]
    close: number[]
    volume: number[]
  }
  moving_averages: {
    sma_20: number[]
    sma_50: number[]
    sma_200: number[]
  }
  bollinger_bands: {
    upper: number[]
    middle: number[]
    lower: number[]
  }
  support_resistance: {
    support_levels: number[]
    resistance_levels: number[]
    pivot: number
  }
  indicators: {
    rsi: number[]
    macd: number[]
    macd_signal: number[]
    macd_histogram: number[]
  }
}

export interface TechnicalAnalysisJob {
  job_id: string
  ticker: string
  status: 'queued' | 'running' | 'completed' | 'failed'
  message: string
  result?: TechnicalAnalysisData
}

File: backend/app/tasks/technical_analysis.py
"""Technical analysis Celery tasks."""

import asyncio
import uuid
from typing import Any

import structlog

from backend.app.celery_app import celery_app

logger = structlog.get_logger(__name__)


def run_async(coro):
    """Run async function in sync context."""
    try:
        loop = asyncio.get_event_loop()
        if loop.is_closed():
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
    except RuntimeError:
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)

    return loop.run_until_complete(coro)


@celery_app.task(bind=True, name="backend.app.tasks.technical_analysis.analyze_stock_technical")
def analyze_stock_technical(
    self,
    ticker: str,
    period: str = "1y",
) -> dict[str, Any]:
    """Perform technical analysis on a stock.

    Args:
        ticker: Stock ticker symbol
        period: Data period (1mo, 3mo, 6mo, 1y, 2y). Defaults to 1y for SMA 200.

    Returns:
        Technical analysis results
    """
    job_id = self.request.id or str(uuid.uuid4())
    ticker = ticker.upper()

    async def run():
        from backend.app.services.yahoo_finance import get_yahoo_finance_client
        from backend.app.services.cache import set_job_progress
        from backend.app.agents.technical_analysis_agent import TechnicalAnalysisAgent

        logger.info("Starting technical analysis", ticker=ticker, job_id=job_id, period=period)

        # Update job status
        await set_job_progress(job_id, "running", 0, "Initializing technical analysis...")

        try:
            result = {"ticker": ticker, "period": period}

            # Step 1: Fetch historical price data (20%)
            await set_job_progress(job_id, "running", 10, "Fetching historical price data...")

            yf_client = get_yahoo_finance_client()

            # Convert period to Yahoo Finance format
            period_map = {
                "1mo": "1mo",
                "3mo": "3mo",
                "6mo": "6mo",
                "1y": "1y",
                "2y": "2y",
            }
            yf_period = period_map.get(period, "6mo")

            prices = await yf_client.get_historical_prices(ticker, period=yf_period, interval="1d")

            if not prices or len(prices) < 50:
                error_msg = f"Insufficient price data for {ticker}. Need at least 50 trading days."
                logger.warning(error_msg, ticker=ticker, data_points=len(prices) if prices else 0)
                await set_job_progress(job_id, "failed", 0, error_msg)
                return {"error": error_msg, "ticker": ticker}

            logger.info("Fetched price data", ticker=ticker, data_points=len(prices))

            # Step 2: Get current price (30%)
            await set_job_progress(job_id, "running", 20, "Fetching current stock info...")

            stock_info = await yf_client.get_stock_info(ticker)
            current_price = stock_info.get("current_price", 0)

            if not current_price:
                # Fallback to latest close price
                current_price = prices[-1].get("close", 0) if prices else 0

            logger.info("Current price", ticker=ticker, price=current_price)

            # Step 3: Run technical analysis (90%)
            await set_job_progress(job_id, "running", 40, "Calculating technical indicators...")

            tech_agent = TechnicalAnalysisAgent()
            analysis_result = await tech_agent.analyze(
                ticker=ticker,
                price_data=prices,
                current_price=current_price
            )

            logger.info("Technical analysis completed",
                       ticker=ticker,
                       signal=analysis_result.overall_signal,
                       composite_score=analysis_result.composite_technical_score)

            # Step 4: Convert dataclass to dict (95%)
            await set_job_progress(job_id, "running", 90, "Preparing results...")

            # Convert result to dict for JSON serialization
            result.update({
                "analysis_date": analysis_result.analysis_date.isoformat(),
                "current_price": analysis_result.current_price,

                # Trend
                "trend_direction": analysis_result.trend.trend_direction,
                "trend_strength_score": analysis_result.trend.trend_strength_score,
                "sma_20": analysis_result.trend.sma_20,
                "sma_50": analysis_result.trend.sma_50,
                "sma_200": analysis_result.trend.sma_200,
                "adx": analysis_result.trend.adx,
                "adx_signal": analysis_result.trend.adx_signal,
                "price_above_sma_20": analysis_result.trend.price_above_sma_20,
                "price_above_sma_50": analysis_result.trend.price_above_sma_50,
                "price_above_sma_200": analysis_result.trend.price_above_sma_200,
                "golden_cross": analysis_result.trend.golden_cross,
                "death_cross": analysis_result.trend.death_cross,

                # Momentum
                "rsi": analysis_result.momentum.rsi,
                "rsi_signal": analysis_result.momentum.rsi_signal,
                "macd": analysis_result.momentum.macd,
                "macd_signal": analysis_result.momentum.macd_signal,
                "macd_histogram": analysis_result.momentum.macd_histogram,
                "macd_cross": analysis_result.momentum.macd_cross,
                "stoch_k": analysis_result.momentum.stoch_k,
                "stoch_d": analysis_result.momentum.stoch_d,
                "stoch_signal": analysis_result.momentum.stoch_signal,
                "roc": analysis_result.momentum.roc,
                "roc_signal": analysis_result.momentum.roc_signal,
                "momentum_score": analysis_result.momentum.momentum_score,

                # Volatility
                "bb_upper": analysis_result.volatility.bb_upper,
                "bb_middle": analysis_result.volatility.bb_middle,
                "bb_lower": analysis_result.volatility.bb_lower,
                "bb_width": analysis_result.volatility.bb_width,
                "bb_signal": analysis_result.volatility.bb_signal,
                "price_position": analysis_result.volatility.price_position,
                "atr": analysis_result.volatility.atr,
                "atr_percent": analysis_result.volatility.atr_percent,
                "volatility_level": analysis_result.volatility.volatility_level,
                "volatility_score": analysis_result.volatility.volatility_score,

                # Volume
                "current_volume": analysis_result.volume.current_volume,
                "avg_volume_20d": analysis_result.volume.avg_volume_20d,
                "volume_ratio": analysis_result.volume.volume_ratio,
                "volume_signal": analysis_result.volume.volume_signal,
                "obv": analysis_result.volume.obv,
                "obv_trend": analysis_result.volume.obv_trend,
                "volume_score": analysis_result.volume.volume_score,

                # Support/Resistance
                "pivot": analysis_result.support_resistance.pivot,
                "resistance_1": analysis_result.support_resistance.resistance_1,
                "resistance_2": analysis_result.support_resistance.resistance_2,
                "resistance_3": analysis_result.support_resistance.resistance_3,
                "support_1": analysis_result.support_resistance.support_1,
                "support_2": analysis_result.support_resistance.support_2,
                "support_3": analysis_result.support_resistance.support_3,
                "support_levels": analysis_result.support_resistance.support_levels,
                "resistance_levels": analysis_result.support_resistance.resistance_levels,
                "nearest_support": analysis_result.support_resistance.nearest_support,
                "nearest_resistance": analysis_result.support_resistance.nearest_resistance,
                "support_distance_pct": analysis_result.support_resistance.support_distance_pct,
                "resistance_distance_pct": analysis_result.support_resistance.resistance_distance_pct,

                # Patterns
                "patterns": analysis_result.patterns.patterns,
                "trend_channel": analysis_result.patterns.trend_channel,
                "consolidation": analysis_result.patterns.consolidation,
                "breakout_signal": analysis_result.patterns.breakout_signal,

                # Overall
                "trend_score": analysis_result.trend_score,
                "price_action_score": analysis_result.price_action_score,
                "composite_technical_score": analysis_result.composite_technical_score,
                "overall_signal": analysis_result.overall_signal,
                "signal_confidence": analysis_result.signal_confidence,

                # Entry Analysis (comprehensive)
                "entry_analysis": {
                    "range_position_pct": analysis_result.entry_analysis.range_position_pct,
                    "range_position_zone": analysis_result.entry_analysis.range_position_zone,
                    "confluence_score": analysis_result.entry_analysis.confluence_score,
                    "confluence_factors": analysis_result.entry_analysis.confluence_factors,
                    "suggested_stop_loss": analysis_result.entry_analysis.suggested_stop_loss,
                    "stop_loss_type": analysis_result.entry_analysis.stop_loss_type,
                    "stop_loss_distance_pct": analysis_result.entry_analysis.stop_loss_distance_pct,
                    "suggested_target": analysis_result.entry_analysis.suggested_target,
                    "target_distance_pct": analysis_result.entry_analysis.target_distance_pct,
                    "risk_reward_ratio": analysis_result.entry_analysis.risk_reward_ratio,
                    "risk_reward_quality": analysis_result.entry_analysis.risk_reward_quality,
                    "is_good_entry": analysis_result.entry_analysis.is_good_entry,
                    "entry_quality": analysis_result.entry_analysis.entry_quality,
                    "entry_quality_score": analysis_result.entry_analysis.entry_quality_score,
                    "suggested_entry_price": analysis_result.entry_analysis.suggested_entry_price,
                    "suggested_entry_zone_low": analysis_result.entry_analysis.suggested_entry_zone_low,
                    "suggested_entry_zone_high": analysis_result.entry_analysis.suggested_entry_zone_high,
                    "wait_for_pullback": analysis_result.entry_analysis.wait_for_pullback,
                    "entry_reasoning": analysis_result.entry_analysis.entry_reasoning,
                    "warning_signals": analysis_result.entry_analysis.warning_signals,
                },

                # Chart data
                "chart_data": analysis_result.chart_data,

                # Data sources
                "data_sources": analysis_result.data_sources,
            })

            # Complete
            await set_job_progress(job_id, "completed", 100, "Technical analysis completed")

            logger.info("Technical analysis task completed",
                       ticker=ticker,
                       job_id=job_id,
                       signal=result["overall_signal"])
            return result

        except Exception as e:
            logger.error("Technical analysis failed", ticker=ticker, error=str(e), exc_info=True)
            error_msg = f"Technical analysis failed: {str(e)}"
            await set_job_progress(job_id, "failed", 0, error_msg)
            raise

    return run_async(run())
